working/augmentation-libraries/feature-specification.md

```diff
@@ -622,311 +662,89 @@ It is a compile-time error if:
-    the other is not. *(Augmentation cannot change `late`-ness, and since being
-    `late` does affect the initializer expression, the augmenting variable is
-    required to repeat the `late`.)*
-
-*   A getter or setter declaration is augmented by an augmenting variable.
-
-*   A late final variable with no initializer expression is augmented by an
-    augmenting variable with an initializer expression.
-    _A late final variable with no initializer has a setter, while one with an
-    initializer does not. An augmentation must not change whether there is a
-    setter._
-
-*   A `const` variable is augmented by an augmenting getter. **(TODO: Can a
-    const variable be augmented by another const variable, changing its value,
-    or is that too weird?)**
-
-*   An `abstract` variable is augmented with a non-abstract variable.
-
-*   An `external` variable is augmented with an `abstract` variable.
-
-### Augmenting enum members
-
-Some enum members can not be augmented: It is a compile-time error if an
-augmenting declaration in an enum declaration (introductory or augmenting)
-has the name `values`, `index`, `hashCode`, or `==`.
-
-*It has always been an error for an enum declaration to declare a member
-named `index`, `hashCode`, `==`, or `values`, and this rule just clarifies
-that this error is applicable for augmenting declarations as well.*
-
-Enum values can _only_ be augmented by enum values, and the implicit getter
-introduced by them is not augmentable. The only thing you are allowed to do
-when augmenting an enum value is add metadata annotations or doc comments.
+*   A function is not complete after all augmentations are applied, unless it
+    is in a context where it can be abstract. *Every function declaration
+    eventually needs to have a body filled in unless it's an instance method
+    that can be abstract. In that case, if no declaration provides a body, it
+    is considered abstract.*
 
-When augmenting an enum value, no constructor invocation should be provided.
-The original value is always used, and the explicit constructor invocation (if
-present) should not be copied.
+### Augmenting variables
 
-New enum values may be defined in an augmenting enum, and they will be appended
-to the current values of the declaration in augmentation application order.
+A class-like augmentation can add *new* variables (that is instance or static
+fields) to the type being augmented:
 
-Augmenting an existing enum value never changes the order in which it appears in
-`values`.
+```dart
+class C {}
 
-For example:
+augment class C {
+  int x = 3;
 
+  static int y = 4;
+}
 ```
-// main.dart
-part 'a.dart';
-part 'c.dart';
-
-enum A {
-  first,
-  second.custom(1);
 
-  final int b;
-
-  const A() : b = 0;
-
-  const A.custom(this.b);
-}
-}
+Variable declarations themselves can't be augmented (by variables, getters, or
+setters), with the exception of abstract instance variable declarations.
```

Comment:
I guess an abstract `augment` getter could be acceptable, and it could be used to add metadata and DartDoc?

Reply:
I considered that, but it gets a little weird. If you do:

```dart
int x;

@someMeta
augment int get x;
```

What does the metadata get added *to*? Metadata is a syntactic construct. It doesn't have semantics at all, so there isn't really an "implicit getter" that it makes sense to add it to. Do we add it to the variable itself? Is that even useful?

I couldn't think of a concrete use case for this so I figured I'd just say that variables are entirely closed to augmentation. We can always loosen this in the future if a compelling use case arises, but that seems unlikely.